connaisseur/sigstore_validator.py

```diff
@@ -0,0 +1,76 @@
+import json
+import logging
+import re
+import subprocess  # nosec
+
+from connaisseur.crypto import decode_and_verify_ecdsa_key
+from connaisseur.exceptions import (
+    CosignError,
+    CosignTimeout,
+    NotFoundException,
+    ValidationError,
+)
+
+
+def get_cosign_validated_digests(image: str, pubkey: str):
+    """
+    Gets and processes cosign validation output for a given `image` and `pubkey`
+    and either returns a list of valid digests or raises a suitable exception
+    in case no valid signature is found or cosign fails.
+    """
+    failed, stdout, stderr = invoke_cosign(image, pubkey)
+    logging.info(
+        "COSIGN output for image: %s; RETURNCODE: %s; STDOUT: %s; STDERR: %s",
+        image,
+        failed,
+        stdout,
+        stderr,
+    )
+    digests = []
+    if not failed:
+        for sig in stdout.splitlines():
+            try:
+                sig_data = json.loads(sig)
+                digests.append(sig_data["Critical"]["Image"]["Docker-manifest-digest"])
+            except json.JSONDecodeError:
+                logging.info("Non-json signature data from Cosign: %s", sig)
+                pass
+    elif stderr == "error: no matching signatures:\nunable to verify signature\n":
+        raise ValidationError(
+            "failed to verify signature of trust data.",
+            {"trust_data_type": "dev.cosignproject.cosign/signature", "stderr": stderr},
+        )
+    elif re.match(r"^error: GET https://[^ ]+ MANIFEST_UNKNOWN:.*", stderr):
+        raise NotFoundException(
+            f'no trust data for image "{image}".',
+            {"trust_data_type": "dev.cosignproject.cosign/signature", "stderr": stderr},
+        )
+    else:
+        raise CosignError(
```

Comment:
For the sake of completeness: If the process dies for other reasons (e.g. there is an issue with the docker image and there is a dysfunctional `cosign` binary) this would result in a `CosignError` which one could argue it not really is. As we also include the error message that could be fine, but I could also image that `CosignError` and `image` could mislead the user in that case. I don't know whether these types could be differentiated at all, but still mentioning ;-)

Reply:
yes, thought about that too, but would not know how to distinguish. Was hoping the error message solves it